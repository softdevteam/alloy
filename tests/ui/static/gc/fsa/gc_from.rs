#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
include!{"./auxiliary/types.rs"}

impl<'a> Drop for HasRef<'a> {
    fn drop(&mut self) {
        use_val(self.a); // should fail
    }
}

fn main() {
    let _: Gc<HasRef> = Gc::from(HasRef::default());
    //~^ ERROR: The drop method for `HasRef<'_>` cannot be safely finalized.
    let _: Gc<HasRef> = Gc::from(Box::new(HasRef::default()));
    //~^ ERROR: The drop method for `HasRef<'_>` cannot be safely finalized.
    let _: Gc<[HasRef]> = Gc::from(vec![HasRef::default()]);
    //~^ ERROR: The drop method for `HasRef<'_>` cannot be safely finalized.
    let _: Gc<[HasRef]> = Gc::from(vec![HasRef::default()].into_boxed_slice());
    //~^ ERROR: The drop method for `HasRef<'_>` cannot be safely finalized.

    // The following should all pass.
    let _: Gc<u8> = Gc::from(1);
    let _: Gc<u8> = Gc::from(Box::new(1));
    let _: Gc<[u8]> = Gc::from(vec![1, 2, 3]);
    let _: Gc<[u8]> = Gc::from(vec![1, 2, 3].into_boxed_slice());
}
