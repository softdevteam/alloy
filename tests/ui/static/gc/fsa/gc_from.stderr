error: The drop method for `HasRef<'_>` cannot be safely finalized.
  --> $DIR/gc_from.rs:13:34
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `&u64`
   |                 because it might not live long enough.
...
LL |     let _: Gc<HasRef> = Gc::from(HasRef::default());
   |                         ---------^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasRef<'_>>` here.
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: The drop method for `HasRef<'_>` cannot be safely finalized.
  --> $DIR/gc_from.rs:15:34
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `&u64`
   |                 because it might not live long enough.
...
LL |     let _: Gc<HasRef> = Gc::from(Box::new(HasRef::default()));
   |                         ---------^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasRef<'_>>` here.
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: The drop method for `HasRef<'_>` cannot be safely finalized.
  --> $DIR/gc_from.rs:17:36
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `&u64`
   |                 because it might not live long enough.
...
LL |     let _: Gc<[HasRef]> = Gc::from(vec![HasRef::default()]);
   |                           ---------^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<[HasRef<'_>]>` here.
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.
   = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

error: The drop method for `HasRef<'_>` cannot be safely finalized.
  --> $DIR/gc_from.rs:19:36
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `&u64`
   |                 because it might not live long enough.
...
LL |     let _: Gc<[HasRef]> = Gc::from(vec![HasRef::default()].into_boxed_slice());
   |                           ---------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<[HasRef<'_>]>` here.
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: aborting due to 4 previous errors

